feature/s3/transfermanager/api_op_GetObject.go

```diff
@@ -0,0 +1,893 @@
+}
+
+// GetObject downloads an object from S3, intelligently splitting large
+// files into smaller parts/ranges according to config and getting them in parallel across
+// multiple goroutines. You can configure the download type, chunk size and concurrency
+// through the Options parameters.
+//
+// Additional functional options can be provided to configure the individual
+// download. These options are copies of the original Options instance, the client of which GetObject is called from.
+// Modifying the options will not impact the original Client and Options instance.
+func (c *Client) GetObject(ctx context.Context, input *GetObjectInput, opts ...func(*Options)) (*GetObjectOutput, error) {
+	i := getter{in: input, options: c.options.Copy()}
+	for _, opt := range opts {
+		opt(&i.options)
+	}
+
+	return i.get(ctx)
+}
+
+type getter struct {
+	options Options
+	in      *GetObjectInput
+	out     *GetObjectOutput
+	w       *types.WriteAtBuffer
+
+	wg sync.WaitGroup
+	m  sync.Mutex
+
+	offset     int64
+	pos        int64
+	totalBytes int64
+	written    int64
+
+	err error
+}
+
+func (g *getter) get(ctx context.Context) (*GetObjectOutput, error) {
+	if err := g.init(ctx); err != nil {
+		return nil, fmt.Errorf("unable to initialize download: %w", err)
+	}
+
+	clientOptions := []func(*s3.Options){
+		func(o *s3.Options) {
+			o.APIOptions = append(o.APIOptions,
+				middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+				addFeatureUserAgent,
+			)
+		}}
+
+	if g.in.PartNumber > 0 {
+		return g.singleDownload(ctx, clientOptions...)
+	}
+
+	var output *GetObjectOutput
+	if g.options.MultipartDownloadType == types.MultipartDownloadTypePart {
+		if g.in.Range != "" {
+			return g.singleDownload(ctx, clientOptions...)
+		}
+		output = g.getChunk(ctx, 1, "", clientOptions...)
+		if g.getErr() != nil {
+			return output, g.err
+		}
+
+		if output.PartsCount > 1 {
+			partSize := output.ContentLength
+			ch := make(chan getChunk, g.options.Concurrency)
+			for i := 0; i < g.options.Concurrency; i++ {
+				g.wg.Add(1)
+				go g.downloadPart(ctx, ch, clientOptions...)
+			}
+
+			for i := int32(2); i <= output.PartsCount; i++ {
+				if g.getErr() != nil {
+					break
+				}
+
+				ch <- getChunk{w: g.w, start: g.pos - g.offset, part: i}
+				g.pos += partSize
+			}
+
+			close(ch)
+			g.wg.Wait()
+		}
+	} else {
+		if g.in.Range == "" {
+			output = g.getChunk(ctx, 0, g.byteRange(), clientOptions...)
+		} else {
+			g.pos, g.totalBytes = g.getDownloadRange()
+			g.offset = g.pos
+		}
+		total := g.totalBytes
+
+		ch := make(chan getChunk, g.options.Concurrency)
+		for i := 0; i < g.options.Concurrency; i++ {
+			g.wg.Add(1)
+			go g.downloadPart(ctx, ch, clientOptions...)
+		}
+
+		// Assign work
+		for g.getErr() == nil {
+			if g.pos >= total {
+				break // We're finished queuing chunks
+			}
+
+			// Queue the next range of bytes to read.
+			ch <- getChunk{w: g.w, start: g.pos - g.offset, withRange: g.byteRange()}
```

Comment:
So as far as I can tell, you're now doing concurrent downloads but you're buffering the entire object in memory before returning it to the caller - which is all well and good until someone goes to try and pull a multi-gigabyte object.

We're basically going to have to return a reader that "composes" the response bodies from inner getPart requests in order. That will allow you to stream the response back to the caller without that extra memory pressure. You might have a "scratch" buffer for that internal composition but naturally it won't be the full object size.